### PR TITLE
Raise clear error for `problem_type="single_label_classification"` with `num_labels=1`

### LINKED ISSUE #45479: `problem_type="single_label_classification"` with `num_labels=1` leads to degenerate zero loss across multiple sequence-classification models
### System Info

Hi, I found what looks like a library-wide issue in `transformers` affecting multiple `ForSequenceClassification` models, not just ModernBERT.

If a model is initialized with:
```python
num_labels=1
problem_type="single_label_classification"
```
the forward pass uses `CrossEntropyLoss()` with only one output logit. This leads to a degenerate zero loss during training/evaluation instead of performing binary classification meaningfully.

I first observed this with `ModernBertForSequenceClassification`, but the same logic appears in other sequence-classification models as well (for example RoBERTa and others using the same loss-selection pattern).

In `modeling_modernbert.py`, the relevant part is:

```python
loss = None
if labels is not None:
    if self.config.problem_type is None:
        if self.num_labels == 1:
            self.config.problem_type = "regression"
        elif self.num_labels > 1 and (labels.dtype == torch.long or labels.dtype == torch.int):
            self.config.problem_type = "single_label_classification"
        else:
            self.config.problem_type = "multi_label_classification"

    if self.config.problem_type == "regression":
        loss_fct = MSELoss()
        if self.num_labels == 1:
            loss = loss_fct(logits.squeeze(), labels.squeeze())
        else:
            loss = loss_fct(logits, labels)
    elif self.config.problem_type == "single_label_classification":
        loss_fct = CrossEntropyLoss()
        loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
    elif self.config.problem_type == "multi_label_classification":
        loss_fct = BCEWithLogitsLoss()
        loss = loss_fct(logits, labels)
```
With `num_labels=1 and problem_type="single_label_classification", this becomes:
```python
CrossEntropyLoss()(logits.view(-1, 1), labels.view(-1))
```

which produces a degenerate zero loss because there is only one class dimension.

Why I think this is a bug:
This setup naturally suggests binary classification with labels like:

- 0 -> class 0
- 1 -> class 1

So from a user perspective, this looks like it should be a valid single-label binary classification setup.
Right now, however, `num_labels=1` is effectively treated as if there were only one possible class in the loss computation, which makes the classification loss meaningless.

### Who can help?

_No response_

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Minimal reproduction
```python
from transformers import AutoModelForSequenceClassification
import torch

model = AutoModelForSequenceClassification.from_pretrained(
    "answerdotai/ModernBERT-base",
    num_labels=1,
    problem_type="single_label_classification",
)

input_ids = torch.tensor([[101, 102]])
attention_mask = torch.tensor([[1, 1]])
labels = torch.tensor([0])

outputs = model(
    input_ids=input_ids,
    attention_mask=attention_mask,
    labels=labels,
)

print(outputs.logits.shape)
print(outputs.loss)
```

Observed result
`outputs.loss` is `0` (or degenerate), and the same behavior also shows up during training.

### Expected behavior

Expected behavior
I would expect `num_labels=1` with `problem_type="single_label_classification"` to support binary classification meaningfully for labels `{0, 1}`, instead of silently producing a degenerate zero loss.
For example, this could be implemented with a single-logit binary objective such as `BCEWithLogitsLoss`, or by internally mapping this configuration to an equivalent binary-classification setup.
In any case, the current behavior of silently returning zero loss seems incorrect.

Actual behavior
The model runs, but training/eval loss becomes degenerate `(0)` because `CrossEntropyLoss` is applied to logits with shape `[..., 1]`.

### PR DESCRIPTION
**Pull request by gaurav0107**
## What

Adds validation in `PreTrainedConfig.__post_init__` that rejects the combination `problem_type="single_label_classification"` + `num_labels=1` with a clear `ValueError` pointing users to the correct setup.

Closes #45479

## Why

Before this change, constructing a sequence-classification model with this combination silently produced a degenerate zero cross-entropy loss (softmax over a single logit is always `1`, so `CrossEntropyLoss` always returns `0`). Training appeared to run but accomplished nothing.

Reproducer (confirmed on `main`, both on shared-loss models like BERT and inlined-loss models like ModernBERT):

```python
import torch
from transformers import BertConfig, BertForSequenceClassification
config = BertConfig(vocab_size=100, hidden_size=32, num_hidden_layers=2,
                    num_attention_heads=2, intermediate_size=64,
                    num_labels=1, problem_type="single_label_classification")
model = BertForSequenceClassification(config)
model.eval()
out = model(input_ids=torch.tensor([[1, 2, 3, 4]]), labels=torch.tensor([0]))
print(out.loss)  # tensor(0.)
```

## Fix

`PreTrainedConfig.__post_init__` now raises `ValueError` on this combination after `num_labels` is resolved from `id2label`. Error message redirects the user to `num_labels=2` (binary classification) or `problem_type="regression"` (single-output regression).

The check sits at the config layer so it covers every sequence-classification model uniformly — both models that delegate to `ForSequenceClassificationLoss` in `src/transformers/loss/loss_utils.py` and models that inline the loss selection (e.g. ModernBERT, ~40 others).

## Coordination / approach

This matches the maintainer direction in [this comment on #45479](https://github.com/huggingface/transformers/issues/45479#issuecomment-4267713872):

> "maybe we could raise a clearer error if users set `num_labels=1` and `problem_type="single_label_classification"` to tell them not to do that?"

## Tests

New regression test `tests/utils/test_configuration_utils.py::ConfigTestUtils::test_single_label_classification_requires_more_than_one_label` asserts the degenerate combination raises, and verifies the three valid shapes (`num_labels=2 + single_label`, `num_labels=1 + regression`, `num_labels=1` with no explicit `problem_type`) still construct.

Verified that the test FAILS on `main` without the fix and PASSES with it.

Ran locally:

```
pytest tests/utils/test_configuration_utils.py
pytest tests/models/bert/test_modeling_bert.py -k "for_sequence or problem_type"
pytest tests/models/modernbert/test_modeling_modernbert.py -k "problem_type or sequence_classification"
pytest tests/models/roberta/test_modeling_roberta.py::RobertaModelTest::test_problem_types
pytest tests/models/albert/test_modeling_albert.py::AlbertModelTest::test_problem_types
```

All pass. `ruff check` and `ruff format --check` clean on the two touched files.

The existing common `test_problem_types` continues to pass: it assigns `problem_type` and `num_labels` as attributes on an already-constructed config, which bypasses `__post_init__` by design, and the test only asserts `loss.backward()` doesn't error.

## Compatibility

- Valid configurations are unaffected.
- Save/load round-trip of valid configs verified.
- A saved Hub config containing this exact bad combination will now fail to load with a clear error. Any such checkpoint can never have been trained successfully (loss was always 0), so no working workflow regresses.

## AI disclosure

AI-assisted: the fix was drafted with Claude, then human-reviewed end-to-end including the reproducer, diff, and test outputs listed above. Disclosed per `CONTRIBUTING.md` "AI-assisted and agentic contributions".

### CODE REVIEW

File: src/transformers/configuration_utils.py [hunk 1]
```diff
@@ -265,6 +265,19 @@ def __post_init__(self, **kwargs):
...
+        # clear message pointing users to the intended setup (use `num_labels=2` for binary
+        # classification, or `problem_type="regression"` for a single-output regression head).
+        # See https://github.com/huggingface/transformers/issues/45479.
+        if self.problem_type == "single_label_classification" and self.num_labels == 1:
+            raise ValueError(
+                '`problem_type="single_label_classification"` requires `num_labels > 1`. With '
+                "`num_labels=1` the cross-entropy loss is degenerate and always zero. For binary "
+                'classification use `num_labels=2`, or use `problem_type="regression"` for a '
+                "single-output regression head."
+            )
```
Comment (Rocketknight1) [inline src/transformers/configuration_utils.py]:
```suggestion
        if self.problem_type == "single_label_classification" and self.num_labels == 1:
            raise ValueError(
                '`problem_type="single_label_classification"` requires `num_labels > 1`. For binary "
                'classification use `num_labels=2`, or use `problem_type="regression"` for a '
                "single-output regression head."
            )
```

File: tests/utils/test_configuration_utils.py [hunk 2]
```diff
@@ -261,6 +261,20 @@ def test_loading_config_do_not_raise_future_warnings(self):
...
+        degenerate zero cross-entropy loss. It must now raise a clear error at config construction.
+        """
+        with self.assertRaises(ValueError):
+            BertConfig(num_labels=1, problem_type="single_label_classification")
+
+        # Valid combinations must still work.
+        BertConfig(num_labels=2, problem_type="single_label_classification")
+        BertConfig(num_labels=1, problem_type="regression")
+        BertConfig(num_labels=1)  # problem_type left unset is fine; it is inferred at forward time.
+
```
Comment (Rocketknight1) [inline tests/utils/test_configuration_utils.py]:
```suggestion
```

**Review by Rocketknight1 — APPROVED** [review]
Looks good, but the code agent was a little verbose! I'll trim it down

File: src/transformers/configuration_utils.py [hunk 3]
```diff
@@ -265,6 +265,13 @@ def __post_init__(self, **kwargs):
             # Keys are always strings in JSON so convert ids to int
             self.id2label = {int(key): value for key, value in self.id2label.items()}
 
+        if self.problem_type == "single_label_classification" and self.num_labels == 1:
+            raise ValueError(
+                '`problem_type="single_label_classification"` requires `num_labels > 1`. For binary "
+                'classification use `num_labels=2`, or use `problem_type="regression"` for a '
+                "single-output regression head."
+            )
```
Comment (Rocketknight1) [inline src/transformers/configuration_utils.py:273]:
```suggestion
        if self.problem_type == "single_label_classification" and self.num_labels == 1:
            raise ValueError(
                '`problem_type="single_label_classification"` requires `num_labels > 1`. For binary '
                'classification use `num_labels=2`, or use `problem_type="regression"` for a '
                "single-output regression head."
            )
```

**Review by Rocketknight1 — APPROVED** [review]

**Comment by gaurav0107** [comment]
Hi @Rocketknight1 
would you mind taking a look at the CI results and helping me understand what caused the failure? The tests_processors job is reporting a worker crash (crashed and worker restarting disabled, exit status 1)

**Comment by Rocketknight1** [comment]
@gaurav0107 probably just an intermittent CI error, I'll throw it into the queue again
